File: FlutterCode/lib/domain/entities/transaction.dart
import 'package:uuid/v8.dart';

enum TransactionType { income, expense }

class Transaction {
  final String id;
  final TransactionType type;
  final double amount;
  final DateTime transactedOn;
  final String title;
  final String notes;
  final String createdByID;
  final DateTime createdOn;
  final String familyID;
  final String categoryID;

  Transaction._({
    required this.id,
    required this.type,
    required this.amount,
    required this.transactedOn,
    required this.title,
    required this.notes,
    required this.createdByID,
    required this.createdOn,
    required this.familyID,
    required this.categoryID,
  });

  static Transaction create({
    required TransactionType type,
    required double amount,
    required DateTime transactedOn,
    required String title,
    String notes = "",
    required String createdByID,
    required String familyID,
    required String categoryID,
  }) {
    if (title.trim().isEmpty) {
      throw ArgumentError("Transaction title cannot be empty.");
    }
    if (amount <= 0) {
      throw ArgumentError("Amount must be greater than zero.");
    }

    final dateOnly = DateTime(
      transactedOn.year,
      transactedOn.month,
      transactedOn.day,
    );

    return Transaction._(
      id: const UuidV8().generate(),
      type: type,
      amount: amount,
      transactedOn: dateOnly,
      title: title,
      notes: notes,
      createdByID: createdByID,
      createdOn: DateTime.now().toUtc(),
      familyID: familyID,
      categoryID: categoryID,
    );
  }

  Transaction rename(String newTitle) {
    if (newTitle.trim().isEmpty) {
      throw ArgumentError("Transaction title cannot be empty.");
    }
    return copyWith(title: newTitle);
  }

  Transaction changeNotes(String newNotes) {
    return copyWith(notes: newNotes);
  }

  Transaction changeAmount(double newAmount) {
    if (newAmount <= 0) {
      throw ArgumentError("Amount must be greater than zero.");
    }
    return copyWith(amount: newAmount);
  }

  Transaction moveToCategory(String newCategoryId) {
    if (newCategoryId.trim().isEmpty) {
      throw ArgumentError("Category cannot be empty.");
    }
    return copyWith(categoryID: newCategoryId);
  }

  Transaction reschedule(DateTime newDate) {
    final dateOnly = DateTime(newDate.year, newDate.month, newDate.day);
    return copyWith(transactedOn: dateOnly);
  }

  Transaction markAsIncome() => copyWith(type: TransactionType.income);
  Transaction markAsExpense() => copyWith(type: TransactionType.expense);

  Transaction copyWith({
    TransactionType? type,
    double? amount,
    DateTime? transactedOn,
    String? title,
    String? notes,
    String? categoryID,
  }) {
    return Transaction._(
      id: id,
      type: type ?? this.type,
      amount: amount ?? this.amount,
      transactedOn: transactedOn ?? this.transactedOn,
      title: title ?? this.title,
      notes: notes ?? this.notes,
      createdByID: createdByID,
      createdOn: createdOn,
      familyID: familyID,
      categoryID: categoryID ?? this.categoryID,
    );
  }

  static Transaction fromJson(Map<String, dynamic> json) {
    return Transaction._(
      id: json['id'],
      type: TransactionType.values.byName(json['type']),
      amount: (json['amount'] as num).toDouble(),
      transactedOn: DateTime.parse(json['transactedOn']),
      title: json['title'],
      notes: json['notes'] ?? "",
      createdByID: json['createdByID'],
      createdOn: DateTime.parse(json['createdOn']),
      familyID: json['familyID'],
      categoryID: json['categoryID'],
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'type': type.name,
        'amount': amount,
        'transactedOn': _formatDateOnly(transactedOn),
        'title': title,
        'notes': notes,
        'createdByID': createdByID,
        'createdOn': createdOn.toIso8601String(),
        'familyID': familyID,
        'categoryID': categoryID,
      };

  static String _formatDateOnly(DateTime d) =>
      "${d.year.toString().padLeft(4, '0')}"
      "-${d.month.toString().padLeft(2, '0')}"
      "-${d.day.toString().padLeft(2, '0')}";
}

File: backend/Expense_Tracker.Domain/TransactionFolder/Transaction.cs
using ErrorOr;
using Expense_Tracker.Domain.CategoryFolder;
using Expense_Tracker.Domain.Common;
using Expense_Tracker.Domain.Errors;
using Expense_Tracker.Domain.FamilyFolder;
using Expense_Tracker.Domain.TransactionFolder.Enums;
using Expense_Tracker.Domain.Users;

namespace Expense_Tracker.Domain.TransactionFolder;

public sealed class Transaction : Entity, IAuditable
{
    public TransactionType Type { get; private set; }
    public decimal Amount { get; private set; }
    public DateOnly TransactedOn { get; private set; }
    public string Title { get; private set; } = string.Empty;
    public string Notes { get; private set; } = string.Empty;

    public Guid CreatedById { get; private set; }
    public Guid FamilyId { get; private set; }
    public Guid CategoryId { get; private set; }

    public User? CreatedBy { get; private set; }
    public Family? Family { get; private set; }
    public Category? Category { get; private set; }

    public DateTimeOffset CreatedAtUtc { get; private set; } = DateTimeOffset.UtcNow;
    public Guid CreatedBy_AuditId { get; private set; } = Guid.Empty;
    public DateTimeOffset LastModifiedUtc { get; private set; } = DateTimeOffset.UtcNow;
    public Guid LastModifiedBy { get; private set; } = Guid.Empty;

    DateTimeOffset ICreatable.CreatedAtUtc
    {
        get => CreatedAtUtc;
        set => CreatedAtUtc = value;
    }

    Guid ICreatable.CreatedBy
    {
        get => CreatedBy_AuditId;
        set => CreatedBy_AuditId = value;
    }

    DateTimeOffset IUpdatable.LastModifiedUtc
    {
        get => LastModifiedUtc;
        set => LastModifiedUtc = value;
    }

    Guid IUpdatable.LastModifiedBy
    {
        get => LastModifiedBy;
        set => LastModifiedBy = value;
    }

    private Transaction() { }

    private Transaction(
        Guid id,
        TransactionType type,
        decimal amount,
        DateOnly transactedOn,
        string title,
        string notes,
        Guid createdByID,
        Guid familyID,
        Guid categoryID) : base(id)
    {
        Type = type;
        Amount = amount;
        TransactedOn = transactedOn;
        Title = title;
        Notes = notes;
        CreatedById = createdByID;
        FamilyId = familyID;
        CategoryId = categoryID;
        CreatedBy_AuditId = createdByID;
    }

    public static ErrorOr<Transaction> Create(
        TransactionType type,
        decimal amount,
        DateOnly transactedOn,
        string title,
        string notes,
        Guid createdByID,
        Guid familyID,
        Guid categoryID)
    {
        if (string.IsNullOrWhiteSpace(title))
            return DomainErrors.GeneralErrors.InvalidState(nameof(Transaction), "Transaction title cannot be empty.");

        if (amount <= 0)
            return DomainErrors.GeneralErrors.InvalidState(nameof(Transaction), "Amount must be greater than zero.");

        if (createdByID == Guid.Empty)
            return DomainErrors.GeneralErrors.InvalidState(nameof(Transaction), "CreatedBy ID cannot be empty.");

        if (familyID == Guid.Empty)
            return DomainErrors.GeneralErrors.InvalidState(nameof(Transaction), "Family ID cannot be empty.");

        if (categoryID == Guid.Empty)
            return DomainErrors.GeneralErrors.InvalidState(nameof(Transaction), "Category ID cannot be empty.");

        var transaction = new Transaction(
            Guid.CreateVersion7(),
            type,
            amount,
            transactedOn,
            title.Trim(),
            notes ?? string.Empty,
            createdByID,
            familyID,
            categoryID);

        return transaction;
    }

    public ErrorOr<Success> Rename(string newTitle)
    {
        if (string.IsNullOrWhiteSpace(newTitle))
            return DomainErrors.GeneralErrors.InvalidState(nameof(Transaction), "Transaction title cannot be empty.");

        Title = newTitle.Trim();
        return new Success();
    }

    public ErrorOr<Success> ChangeNotes(string newNotes)
    {
        Notes = newNotes ?? string.Empty;
        return new Success();
    }

    public ErrorOr<Success> ChangeAmount(decimal newAmount)
    {
        if (newAmount <= 0)
            return DomainErrors.GeneralErrors.InvalidState(nameof(Transaction), "Amount must be greater than zero.");

        Amount = newAmount;
        return new Success();
    }

    public ErrorOr<Success> MoveToCategory(Guid newCategoryId)
    {
        if (newCategoryId == Guid.Empty)
            return DomainErrors.GeneralErrors.InvalidState(nameof(Transaction), "Category ID cannot be empty.");

        CategoryId = newCategoryId;
        return new Success();
    }

    public ErrorOr<Success> Reschedule(DateOnly newDate)
    {
        TransactedOn = newDate;
        return new Success();
    }

    public ErrorOr<Success> MarkAsIncome()
    {
        Type = TransactionType.Income;
        return new Success();
    }

    public ErrorOr<Success> MarkAsExpense()
    {
        Type = TransactionType.Expense;
        return new Success();
    }
}
